aten/src/ATen/native/native_functions.yaml

```diff
@@ -6472,6 +6472,34 @@
     SparseCPU, SparseCUDA: norm_sparse
   autogen: native_norm.ScalarOpt_dim_dtype_out
 
+# return values are: output, save_mean, save_var, reserve
+# reserve tensor is only used in cudnn
+- func: batch_norm_with_update(Tensor input, Tensor? weight, Tensor? bias, Tensor(a!) running_mean, Tensor(b!) running_var, float momentum, float eps, bool cudnn_enabled) -> (Tensor, Tensor, Tensor, Tensor)
+  dispatch:
+    CPU: batch_norm_with_update_cpu
+    CUDA: batch_norm_with_update_cuda
+    MPS: batch_norm_with_update_mps
+    MkldnnCPU: batch_norm_with_update_mkldnn
+  autogen: batch_norm_with_update_functional
+
+- func: batch_norm_no_update(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, float momentum, float eps, bool cudnn_enabled) -> (Tensor, Tensor, Tensor, Tensor)
+  dispatch:
+    CompositeExplicitAutograd: batch_norm_no_update
+  autogen: batch_norm_no_update.out
+
+- func: batch_norm_with_update.out(Tensor input, Tensor? weight, Tensor? bias, Tensor(a!) running_mean, Tensor(b!) running_var, float momentum, float eps, bool cudnn_enabled, *, Tensor(d!) out, Tensor(e!) save_mean, Tensor(f!) save_invstd, Tensor(g!) reserve) -> (Tensor(d!), Tensor(e!), Tensor(f!), Tensor(g!))
+  dispatch:
+    CPU: batch_norm_with_update_cpu_out
+    CUDA: batch_norm_with_update_cuda_out
+    MPS: batch_norm_with_update_mps_out
+
+- func: batch_norm_backward(Tensor grad_output, Tensor input, Tensor weight, Tensor? running_mean, Tensor? running_var, Tensor? save_mean, Tensor? save_var, bool update, float eps, bool[3] grad_input_mask, Tensor reserve, bool cudnn_enabled) -> (Tensor, Tensor, Tensor)
```

Comment:
Is it best to share a complex backward op here or have one for no_update and one for with_update? cc @bdhirsh in particular if you prefer to see less optional and things in the graph?

Reply:
Hmm... 3 areas where I think it's worth thinking this through are:

(1) for functionalization, it doesn't matter (regardless of the update / no_update forward, the backward will have no mutation

(2) for torch.compile, it doesn't matter either (inductor will decompose these ops into primitives that it fuses either way, so "the decomp will handle it").

(3) export. You're right and we should probably be explicit about what we want in the backward graph for export. Maybe @tugsbayasgalan has thoughts?

I think my vote would be that.. for export, a single op is possible better, because it minimizes the number of differences in the graph across train() vs. eval()?

Reply:
And separately: what is the argument for having a new backward op again, if all it does is call into the existing `native_batch_norm_backward` op? Naming consistency with the forward? I think that's probably reasonable, just want to make sure we're all in agreement on the new op

Reply:
> And separately: what is the argument for having a new backward op again, if all it does is call into the existing native_batch_norm_backward op?

I think one argument here is eventually we want to remove the ops in the old stack. This includes `native_batch_norm` and `native_batch_norm_backward`. If we don't add a new backward op here then it'll be harder to do that

Reply:
Thanks for the detailed explanation Brian.
I'm fine with the single op then!
And yes let's use the new name so that we can delete all the old stuff :D